qtism/runtime/tests/AbstractSessionManager.php

```diff
@@ -247,8 +247,10 @@ protected function createRoute(AssessmentTest $test)
                             $route->getLastRouteItem()->addBranchRules($current->getBranchRules());
 
                             // Do the same as for branch rules for pre conditions, except that they must be
-                            // attached on the first item of the route.
-                            $route->getFirstRouteItem()->addPreConditions($current->getPreConditions());
+                            // attached on all the items of the selected route.
+                            foreach ($route as $selectableRouteItem) {
+                                $selectableRouteItem->addPreConditions($current->getPreConditions());
+                            }
```

Comment:
Hey, dude, I have small proposal:
what if we add `hasPreConditions` method in both `TestPart` and `SectionPart` classes (though we only need `SectionPart` changes for the current case)? The `foreach` exporession would then be changed to smth like 
```php
if ($current->hasPreConditions()) {
    foreach ($route as $selectableRouteItem) {
        ...
    }
}
```

It would make "our" lifes a bit easier, meaning that we won't need to go through _each existing_ item in _each existing_ test sections, even if we don't have any preconditions.
How does that sound for you?

Reply:
A good suggestion in general, but it won't really make any difference in terms of complexity or performance. In fact it will increase the cyclomatic complexity in the end.

Reply:
I know, I was just testing it in local, and was a bit tired to go through this line on big sections with no preconditions and a lot of items while debugging. that's why decided to propose :)

Reply:
I will follow @wazelin on this one. Tend to agree. Thanks @uncleempty for the proposal!